include/Gaffer/Private/IECorePreview/TaskMutex.h

```diff
@@ -0,0 +1,357 @@
+					ExecutionStateMutex::scoped_lock executionStateLock( m_mutex->m_executionStateMutex );
+					assert( !m_mutex->m_executionState );
+					m_mutex->m_executionState = std::make_shared<ExecutionState>();
+					executionStateLock.release();
+
+					m_mutex->m_executionState->arena.execute(
+						[this, &f] {
+							// Note : We deliberately call `run()` and `wait()` separately
+							// instead of calling `run_and_wait()`. The latter is buggy until
+							// TBB 2018 Update 3, causing calls to `wait()` on other threads to
+							// return immediately rather than do the work we want.
+							m_mutex->m_executionState->taskGroup.run( f );
+							m_mutex->m_executionState->taskGroup.wait();
+						}
+					);
+
+					executionStateLock.acquire( m_mutex->m_executionStateMutex );
+					m_mutex->m_executionState = nullptr;
+				}
+
+				/// Acquires mutex or returns false. Never does TBB tasks.
+				bool tryAcquire( TaskMutex &mutex, bool write = true )
+				{
+					return acquireOr( mutex, write, []( bool workAvailable ){ return false; } );
+				}
+
+				/// Releases the lock. This will be done automatically
+				/// by ~ScopedLock, but may be called explicitly to release
+				/// the lock early.
+				void release()
+				{
+					assert( m_mutex );
+					if( !m_recursive )
+					{
+						m_lock.release();
+					}
+					m_mutex = nullptr;
+				}
+
+				/// Advanced API
+				/// ============
+				///
+				/// These methods provide advanced usage required by complex requirements
+				/// in Gaffer's LRUCache. They should not be considered part of the canonical
+				/// API.
+
+				/// Returns true if `acquire()` obtained a recursive lock rather
+				/// than a unique lock. Recursive locks are available to any thread
+				/// performing work on behalf of `execute()`.
+				bool recursive() const
+				{
+					return m_recursive;
+				}
+
+				/// Tries to acquire the mutex, returning true on success. On failure,
+				/// calls `workNotifier( bool workAvailable )`. If work is available and
+				/// `workNotifier` returns true, then this thread will perform TBB tasks
+				/// spawned by `execute()` until the work is complete. Returns false on
+				/// failure regardless of whether or not work is done.
+				template<typename WorkNotifier>
+				bool acquireOr( TaskMutex &mutex, bool write, WorkNotifier &&workNotifier )
+				{
+					assert( !m_mutex );
+					if( m_lock.try_acquire( mutex.m_mutex, write ) )
+					{
+						// Success!
+						m_mutex = &mutex;
+						m_recursive = false;
+						m_writer = write;
+						return true;
+					}
+
+					ExecutionStateMutex::scoped_lock executionStateLock( mutex.m_executionStateMutex );
+					if( mutex.m_executionState && mutex.m_executionState->arenaObserver.containsThisThread() )
+					{
+						m_mutex = &mutex;
+						m_writer = false;
+						m_recursive = true;
+						return true;
+					}
+
+					const bool workAvailable = mutex.m_executionState.get();
+					if( !workNotifier( workAvailable ) || !workAvailable )
```

Comment:
Potentially a silly Q, just trying to get my head around the significance of calling workNotifier function even if there is no work available, as in cases such as at :143 this would return true regardless.

Reply:
Good question. I had originally written it with a `workAcceptor`, called only when work was available, and then rewrote it to be like this. My thinking was as follows :

- I felt that `acquireOr( mutex, someFunction )` strongly implied that you would either acquire a lock on the mutex or `someFunction()` would be called. Having a third case where basically nothing happened seemed surprising, and less self-documenting.
- The motivating use case for `acquireOr()` is more straightforwardly written with the new behaviour. This'll be in an upcoming PR, but basically a revamped LRUCache needs to move `[this block](
https://github.com/ImageEngine/cortex/blob/master/include/IECore/LRUCache.inl#L442-L449)`  into a `workNotifier` so that the bin lock is released _before_ the work is done. With merely a `workAcceptor`, it would need more state to track whether work was done, and then put a conditional `if( !workDone ) { binLock.release(); }` after the return from `acquireOr()`. It can be done, but the logic in there is already quite complex, and I felt the `workNotifier` version was easier to follow.

It's true that the return value is ignored if work wasn't available, but I felt that on balance the points above outweighed that.